### PR TITLE
Card (264-FO-saisir/292-FO-ergonomie)

### PR DESCRIPTION
**Pull request by jalelmahjoubi**

### CODE REVIEW

File: fix.patch [hunk 1]
```diff
@@ -0,0 +1,337 @@
+From 1921f06af231bfe45161f99b62a9edf510161073 Mon Sep 17 00:00:00 2001
```
Comment (EmmanuelDemey) [inline fix.patch]: You should not push this file.

File: src/constants/dictionary.js [hunk 2]
```diff
@@ -332,6 +343,42 @@ const dictionary = {
...
+  }, 
+
+  months: {
+    en: 'months',
+    fr: 'mois',
+  }, 
+
+  dateinitial: {
+    en: 'Date format',
+    fr: 'Date au format',
```
Comment (EmmanuelDemey) [inline src/constants/dictionary.js]:
```suggestion
    fr: 'Format de la date',
```

à confirmer

File: src/constants/dictionary.js [hunk 3]
```diff
@@ -340,6 +387,12 @@ const dictionary = {
     en: 'Select an unit of measure',
     fr: 'Sélectionnez une unité de mesure',
   },
+
+  dateEmptySelect: {
+    en: 'Select a date format',
+    fr: 'Sélectionnez une format de date',
```
Comment (EmmanuelDemey) [inline src/constants/dictionary.js]:
```suggestion
    fr: 'Sélectionnez un format de date',
```

File: src/constants/dictionary.js [hunk 4]
```diff
@@ -758,6 +811,31 @@ const dictionary = {
     fr: 'Obligatoire',
     en: 'Mandatory',
   },
+
+  durationformat1: {
+    fr: 'heures/minutes',
+    en: 'hours/minutes',
```
Comment (EmmanuelDemey) [inline src/constants/dictionary.js]:
```suggestion
    en: 'hours/minuts',
```

File: src/forms/validation-rules.js [hunk 5]
```diff
@@ -195,7 +195,10 @@ export function validCollectedVariables(
    * are in the right order.
    */
   const isTheSameOrder = true;
-
+  /*expectedVariables[0].name &&
```
Comment (EmmanuelDemey) [inline src/forms/validation-rules.js]: Could you remove this commented code ?

File: src/model/transformations/external-variable.js [hunk 6]
```diff
@@ -37,15 +41,17 @@ export function remoteToStore(remote = []) {
...
       },
     };
   }, {});
 }
 
 export function storeToRemote(store) {
+  
   return Object.keys(store).map(key => {
-    const {
+    let {
```
Comment (EmmanuelDemey) [inline src/model/transformations/external-variable.js]: Why have you changed const to let ?

File: src/model/transformations/response-format-simple.js [hunk 7]
```diff
@@ -12,21 +13,50 @@ export function remoteToState(remote) {
...
   if (unit !== undefined) datatype.unit = unit;
+  if (format !== undefined) datatype.format = format;
+
+  if (datatype.minimum !== undefined && typeName==='DURATION' ) {
+    let strminimum = datatype.minimum;
+    let strmaximum = datatype.maximum;
+    let matches_minimum = strminimum.match(/\d+/g);
+    let matches_maximum = strmaximum.match(/\d+/g);
+    if (format !== undefined && format === "PTnHnM" ) datatype.mihours = matches_minimum[0];
+    if (format !== undefined && format === "PTnHnM" ) datatype.miminutes =  matches_minimum[1];
```
Comment (EmmanuelDemey) [inline src/model/transformations/response-format-simple.js]: Why not setting this value in the same if block ?
```
if (format !== undefined && format === "PTnHnM" ) {
     datatype.mihours = matches_minimum[0];
     datatype.miminutes =  matches_minimum[1];
}
```

File: src/model/transformations/response-format-table.js [hunk 8]
```diff
@@ -74,12 +74,42 @@ function getResponsesOffset(primaryState, secondaryState, activeCodeLists) {
 function getMeasuresModel(responses, dimensions, offset) {
   const responsesModel = [];
   for (let i = 0; i < dimensions.length; i += 1) {
+
+    if (responses[i].Datatype.Minimum !== undefined && responses[i].Datatype.typeName==='DURATION' ) {
+      let strminimum = responses[i].Datatype.Minimum;
+      let strmaximum = responses[i].Datatype.Maximum;
+      let matches_minimum = strminimum.match(/\d+/g);
+      let matches_maximum = strmaximum.match(/\d+/g);
+      if (responses[i].Datatype.Format !== undefined && responses[i].Datatype.Format === "PTnHnM" ) responses[i].Datatype.Mihours = matches_minimum[0];
```
Comment (EmmanuelDemey) [inline src/model/transformations/response-format-table.js]: The same remark as previously

File: src/utils/validation/validation-rules.js [hunk 9]
```diff
@@ -201,12 +208,12 @@ export const tableListMeasuresRules = {
   [`${RESPONSE_FORMAT}.${TABLE}.${LIST_MEASURE}.${SINGLE_CHOICE}.${DEFAULT_CODES_LIST_SELECTOR_PATH}`]: [
     validCodesList,
   ],
-  [`${RESPONSE_FORMAT}.${TABLE}.${LIST_MEASURE}.${SIMPLE}.${NUMERIC}.minimum`]: [
-    value => minValue(0)(value),
-  ],
-  [`${RESPONSE_FORMAT}.${TABLE}.${LIST_MEASURE}.${SIMPLE}.${NUMERIC}.maximum`]: [
-    value => minValue(1)(value),
-  ],
+  // [`${RESPONSE_FORMAT}.${TABLE}.${LIST_MEASURE}.${SIMPLE}.${NUMERIC}.minimum`]: [
```
Comment (EmmanuelDemey) [inline src/utils/validation/validation-rules.js]: You should remove this commented code

File: src/widgets/component-new-edit/components/response-format/simple/response-format-simple.jsx [hunk 10]
```diff
@@ -58,11 +60,16 @@ class ResponseFormatSimple extends Component {
           <View key={TEXT} value={TEXT} label={Dictionary.TEXT}>
             <ResponseFormatDatatypeText />
           </View>
-          <View key={DATE} value={DATE} label={Dictionary.DATE} />
+          <View key={DATE} value={DATE} label={Dictionary.DATE} >
+          <ResponseFormatDatatypeDate />
```
Comment (EmmanuelDemey) [inline src/widgets/component-new-edit/components/response-format/simple/response-format-simple.jsx]: Do you format your code ? you should use the built-in prettier formatting tool we have on this project.

File: src/model/transformations/response-format-simple.js [hunk 11]
```diff
@@ -35,10 +65,58 @@ export function remoteToState(remote) {
...
+
+
+    
+  if (typeName === 'DURATION' && dataType.format === 'PnYnM' ){
+    dataType.miyears==''?dataType.miyears = 0: dataType.miyears;
+    dataType.mimonths==''?dataType.mimonths = 0: dataType.mimonths;
+    dataType.mayears==''?dataType.mayears = 0: dataType.mayears;
+    dataType.mamonths==''?dataType.mamonths = 0: dataType.mamonths;
+ 
+    dataType.minimum = "P"+dataType.miyears+"Y"+dataType.mimonths+"M";
```
Comment (EmmanuelDemey) [inline src/model/transformations/response-format-simple.js]: You should use string interpolation
```suggestion
    dataType.minimum = `P${dataType.miyears}Y${dataType.mimonths}M`;
```

File: src/model/transformations/response-format-simple.js [hunk 12]
```diff
@@ -35,10 +65,58 @@ export function remoteToState(remote) {
...
+
+    if(dataType.maximum ===  ''){
+      delete dataType.maximum;
+    }
+  }
+
+
+    
+  if (typeName === 'DURATION' && dataType.format === 'PnYnM' ){
+    dataType.miyears==''?dataType.miyears = 0: dataType.miyears;
```
Comment (EmmanuelDemey) [inline src/model/transformations/response-format-simple.js]: I will prefer using this syntax for ternary statement
```suggestion
    dataType.miyears = dataType.miyears === '' ? 0 : dataType.miyears;
```

File: src/model/transformations/response-format-simple.js [hunk 13]
```diff
@@ -35,10 +65,58 @@ export function remoteToState(remote) {
...
+    dataType.mayears==''?dataType.mayears = 0: dataType.mayears;
+    dataType.mamonths==''?dataType.mamonths = 0: dataType.mamonths;
+ 
+    dataType.minimum = "P"+dataType.miyears+"Y"+dataType.mimonths+"M";
+    dataType.maximum = "P"+dataType.mayears+"Y"+dataType.mamonths+"M";
+
+  } 
+
+  if (typeName === 'DURATION' && dataType.format === 'PTnHnM' ){
+    dataType.mihours==''?dataType.mihours = 0: dataType.mihours;
```
Comment (EmmanuelDemey) [inline src/model/transformations/response-format-simple.js]: SAme remark as previously

File: src/model/transformations/response-format-simple.js [hunk 14]
```diff
@@ -35,10 +65,58 @@ export function remoteToState(remote) {
...
+    dataType.miminutes==''?dataType.miminutes = 0: dataType.miminutes;
+    dataType.mahours==''?dataType.mahours = 0: dataType.mahours;
+    dataType.maminutes==''?dataType.maminutes = 0: dataType.maminutes;
+    dataType.minimum = "PT"+dataType.mihours+"H"+dataType.miminutes+"M";
+    dataType.maximum = "PT"+dataType.mahours+"H"+dataType.maminutes+"M";
+  } 
+  
+  if(typeName === 'DURATION'){
+    delete dataType.miyears;
```
Comment (EmmanuelDemey) [inline src/model/transformations/response-format-simple.js]: Can you avoid using the delete keyword ?

File: src/model/transformations/response-format-table.js [hunk 15]
```diff
@@ -74,12 +74,42 @@ function getResponsesOffset(primaryState, secondaryState, activeCodeLists) {
...
+      if (responses[i].Datatype.Format !== undefined && responses[i].Datatype.Format === "PTnHnM" ) responses[i].Datatype.Miminutes =  matches_minimum[1];
+      if (responses[i].Datatype.Format !== undefined && responses[i].Datatype.Format === "PTnHnM" ) responses[i].Datatype.Mahours = matches_maximum[0];
+      if (responses[i].Datatype.Format !== undefined && responses[i].Datatype.Format === "PTnHnM" ) responses[i].Datatype.Maminutes = matches_maximum[1];
+      if (responses[i].Datatype.Format !== undefined && responses[i].Datatype.Format === "PnYnM" ) responses[i].Datatype.Miyears = matches_minimum[0];
+      if (responses[i].Datatype.Format !== undefined && responses[i].Datatype.Format === "PnYnM" ) responses[i].Datatype.Mimonths = matches_minimum[1];
+      if (responses[i].Datatype.Format !== undefined && responses[i].Datatype.Format === "PnYnM" ) responses[i].Datatype.Mayears = matches_maximum[0];
+      if (responses[i].Datatype.Format !== undefined && responses[i].Datatype.Format === "PnYnM" ) responses[i].Datatype.Mamonths = matches_maximum[1];
+
+      responses[i].Datatype.Mihours == 0? responses[i].Datatype.Mihours ='': responses[i].Datatype.Mihours;
```
Comment (EmmanuelDemey) [inline src/model/transformations/response-format-table.js]: Same remark as previously

File: src/model/transformations/response-format-table.js [hunk 16]
```diff
@@ -233,6 +264,39 @@ function stateToResponseState(state) {
     };
   }
 
+  if (responseState.typeName === 'DURATION' && responseState.format === 'PnYnM' ){
+
+    responseState.miyears==''?responseState.miyears= 0: responseState.miyears;
```
Comment (EmmanuelDemey) [inline src/model/transformations/response-format-table.js]: Same remark as previously

File: src/model/transformations/response-format-table.js [hunk 17]
```diff
@@ -233,6 +264,39 @@ function stateToResponseState(state) {
...
+    responseState.mayears==''?responseState.mayears = 0: responseState.mayears;
+    responseState.mamonths==''?responseState.mamonths = 0: responseState.mamonths;
+
+    responseState.minimum = "P"+responseState.miyears+"Y"+responseState.mimonths+"M";
+    responseState.maximum = "P"+responseState.mayears+"Y"+responseState.mamonths+"M";
+  } 
+
+  if (responseState.typeName === 'DURATION' && responseState.format === 'PTnHnM' ){
+
+    responseState.mihours==''?responseState.mihours= 0: responseState.mihours;
```
Comment (EmmanuelDemey) [inline src/model/transformations/response-format-table.js]: Same remark as previously

File: src/model/transformations/response-format-table.js [hunk 18]
```diff
@@ -233,6 +264,39 @@ function stateToResponseState(state) {
...
+    responseState.miminutes==''?responseState.miminutes = 0: responseState.miminutes;
+    responseState.mahours==''?responseState.mahours = 0: responseState.mahours;
+    responseState.maminutes==''?responseState.maminutes = 0: responseState.maminutes;
+
+    responseState.minimum = "PT"+responseState.mihours+"H"+responseState.miminutes+"M";
+    responseState.maximum = "PT"+responseState.mahours+"H"+responseState.maminutes+"M";
+  } 
+
+  if(responseState.typeName === 'DURATION'){
+    delete responseState.miyears;
```
Comment (EmmanuelDemey) [inline src/model/transformations/response-format-table.js]: Can you avoid using the delete keyword ?

File: src/widgets/component-new-edit/components/response-format/simple/response-format-simple.jsx (hunk 10 above)
Comment (jalelmahjoubi) [reply src/widgets/component-new-edit/components/response-format/simple/response-format-simple.jsx]: @EmmanuelDemey  i didnt format my code and i used simple-date.jsx file to insert the new input

File: src/constants/dictionary.js [hunk 19]
```diff
@@ -332,6 +343,42 @@ const dictionary = {
...
+  }, 
+
+  months: {
+    en: 'months',
+    fr: 'mois',
+  }, 
+
+  dateinitial: {
+    en: 'Date format',
+    fr: 'Date de la format',
```
Comment (EmmanuelDemey) [inline src/constants/dictionary.js]: Format de la date

File: src/constants/dictionary.js [hunk 20]
```diff
@@ -332,6 +343,42 @@ const dictionary = {
...
+  },
+
+  durationinitial: {
+    en: 'Duration format',
+    fr: 'Durée au format',
```
Comment (EmmanuelDemey) [inline src/constants/dictionary.js]: Format de la durée

File: src/layout/generic-input/components/generic-input.jsx [hunk 21]
```diff
@@ -95,6 +122,29 @@ class GenericInput extends Component {
...
+         {({ isActive, onCancel, onConfirm }) => {
+           if (isActive) {
+             return (
+               <ReactModal 
+                 isOpen={true}
+                 ariaHideApp={false}
+                 shouldCloseOnOverlayClick={false}
+                 style={customModalStyles}
+                 >
+                 <div>
```
Comment (EmmanuelDemey) [inline src/layout/generic-input/components/generic-input.jsx]: Do you already need this div ?

**Pull request merged**